tools/document-pii/src/item_collector.rs

```diff
@@ -0,0 +1,325 @@
+            vec![quote::quote!(#leading_path).to_string()]
+        }
+        // Currently this script can't handle glob imports, which, we shouldn't use anyway.
+        UseTree::Glob(_) => vec![quote::quote!(#leading_path).to_string()],
+    }
+}
+
+fn crate_name_from_file(file_path: &Path) -> anyhow::Result<String> {
+    let file_str = file_path
+        .as_os_str()
+        .to_str()
+        .ok_or_else(|| anyhow!("Invalid file path: {}", file_path.display()))?;
+
+    let src_index = file_str
+        .find("/src/")
+        .or_else(|| file_str.find("\\src\\"))
+        .ok_or_else(|| {
+            anyhow!(
+                "Invalid file path (missing '/src/' or '\\src\\'): {}",
+                file_path.display()
+            )
+        })?;
+
+    let back_index = file_str[..src_index]
+        .rfind('/')
+        .or_else(|| file_str[..src_index].rfind('\\'))
+        .ok_or_else(|| {
+            anyhow!(
+                "Invalid file path (missing separator before '/src/' or '\\src\\'): {}",
+                file_path.display()
+            )
+        })?
+        + 1;
+
+    Ok(file_str
+        .split_at(src_index)
+        .0
+        .split_at(back_index)
+        .1
+        .replace('-', "_"))
+}
+
+fn add_file_stem_to_module_path(
+    file_path: &Path,
+    module_path: &mut Vec<String>,
+) -> anyhow::Result<()> {
+    let file_stem = file_path
+        .file_stem()
+        .ok_or_else(|| {
+            anyhow!(
+                "Invalid file path (unable to find file stem): {}",
+                file_path.display()
+            )
+        })?
+        .to_string_lossy()
+        .into_owned();
+
+    module_path.push(file_stem);
+    Ok(())
+}
+
+fn split_path_at_crate_name(module_path: &[String], file_path: &Path) -> anyhow::Result<String> {
+    // Build the final use-path by joining the module path components
+    // with "::" and removing the '-' character if any, since Rust module
+    // names replace '-' with '_'.
+    let use_path = module_path.join("::");
+    let crate_name = crate_name_from_file(file_path)?;
+
+    let index = use_path.find(&crate_name).ok_or_else(|| {
+        anyhow!(
+            "Failed to find crate name '{}' in use path '{}'",
+            crate_name,
+            use_path
+        )
+    })?;
+    let use_path = use_path.split_at(index).1.to_string();
+
+    Ok(use_path)
+}
+
+/// Takes in the path to a Rust file and returns the path as you'd refer to it in a use-statement.
+/// e.g. "/Users/tor/prog/rust/relay/relay-general/src/protocol/types.rs" -> "relay_general::protocol"
+fn module_name_from_file(file_path: &Path) -> anyhow::Result<String> {
+    let mut module_path = file_path
+        .parent()
+        .ok_or_else(|| {
+            anyhow!(
+                "Invalid file path (unable to find parent directory): {}",
+                file_path.display()
+            )
+        })?
+        .components()
+        .map(|part| part.as_os_str().to_string_lossy().replace('-', "_"))
```

Comment:
The `replace` called in multiple places on, what it looks like, same things. Maybe some the things, like the file path can be first resolved, then normalized once and you can avoid normalazing same things over and over again?

Reply:
rewrote it a bit now, no longer does the same replace twice